Replace unprincipled "toeless sRGB" gamma function for PS1-derived movies with BT1886 Appendix 1 function with constants selected to match a properly calibrated mid-90s Sony Trinitron CRT.

ChthonVII opened the pull request:
## Summary

(This is a resubmit of pull 814 to fix the damn inconsistent-EOL issues (or, rather, to _not_ change the EOL marks in changed files despite the inconsistency) while keeping a clean commit history.)

Replaces the default gamma function for movies (that lack metadata specifying a particular gamma function). Removes the unprincipled "toeless sRGB" gamma function. Replaces it with BT1886 Appendix 1 function with constants selected to match a properly calibrated mid-90s Sony Trinitron CRT.

(Note that BT1886 specifies _two_ EOTF functions: The standard "Annex 1" function, which is rather useless; and the more accurate "Appendix 1" function, which is used here.)

### Motivation

"Toeless sRGB" was always an unprincipled hack, motivated largely by the unsuitablity of every alternative I knew of at the time.

BT1886 Appendix 1 is correct in principle, because these videos were intended to be played back on a CRT television. (And "mid 90's Sony Trinitron" is a reasonable guess for which CRT television the original developers had on their desks.) In addition to being right in principle, it does everything "toeless sRGB" was meant to -- no banding near black around the stars in the FF7 opening movie, shadowy areas aren't totally lost to blackness, brights aren't blown out.

### ACKs

- [x] I have updated the [Changelog.md](https://github.com/julianxhokaxhiu/FFNx/blob/master/Changelog.md) file
- [x] I did test my code on FF7
- [ ] I did test my code on FF8 -  I don't have FF8 installed, but the behavior should be consistent with FF7.

julianxhokaxhiu commented on Changelog.md:
```diff
@@ -4,6 +4,10 @@
+- Movie: Replace unprincipled "toeless sRGB" gamma function for PS1-derived movies with BT1886 Appendix 1 function with constants selected to match a properly calibrated mid-90s Sony Trinitron CRT.
+
+## Common
+
```
Please compact this section and add the single new change line after the already existing one in the changelog ( top to bottom ). Thank you

ChthonVII commented:
Please don't merge this just yet. A properly calibrated Trinitron is too dark for FF7. I'm working on finding the best looking parameters.

julianxhokaxhiu commented:
Thanks for raising this concern. I want to be very honest on the questions you asked but let me try to answer them separately.

>  Would you rather I space them out into this pull plus two more, or lump them altogether into one big pull?

For the sake of the review, the smallest the code patch, the easier is to review it. Huge bloated PRs tend to usually sit there for weeks until I find the proper time to go through them. I personally prefer to have them packed as a single unit feature, hopefully involving less code patches as possible. If you could follow the same approach I'd appreciate.

> Do we want to make more modes to reproduce what European or American audiences would have seen on their TVs? Or what people would have seen on late-90s CRT computer monitors? Alternatively, do we want to just get rid of uncorrected mode?

So to be very frank, one of the reasons why I'm not able to introduce the hardware acceleration for 4K movies in FFNx is exactly because of this layer that requires massive rework to be included while using the ffmpeg HW acceleration layer ( which happenns all in the GPU ). Instead I'd like to ask you, would it be possible so to say to move all this code and logic into a separate Reshade shader for instance? I've seen other games doing wonder with it on top, and FFNx supports that, would you be open in case to develop a separate solution based on ReShade? I'll be more than happy then to find a way to integrate it maybe as a custom mod in 7th.

Would you be open for it?

ChthonVII commented:
> [W]ould it be possible so to say to move all this code and logic into a separate Reshade shader?

No, I don't believe that would be possible. The sticking point is that the game engine draws other stuff on top of movies. This means that we have to deal with the movie's color properties right then and there, getting it into the same gamut/color space/gamma as the stuff that's about to get drawn on top of it. I don't not see any way around this.

Also, you may recall that we had to take responsibility for movie tv->pc range expansion away from ffmpeg and do it in FFNx's shader because ffmpeg had a bug causing it to apply the range expansion twice in some cases (including SYW's upscales). So far as I know, that bug still isn't fixed.

Another issue is that, for our most common use case for FF7, PS1-derived videos, we want a gamma function that ffmpeg just doesn't have. (Which I am now replacing with a better gamma function that ffmpeg doesn't have.)

Finally, I do not care for reshade one bit and would not be interested in working on in.

So, at any rate, I guess I will take this to mean that you definitely aren't interested in adding more color modes.

As far as simplifying the existing shader code, I can think of two things that would trim some code:
1. Remove the stubs for per-asset color gamut. No one seems interested in getting the metadata handling into 7th Heaven. So it's just dead code at this point.
2. Remove uncorrected mode. I guess we should see how NTSC-J looks after I'm done overhauling it.

julianxhokaxhiu commented:
Yep, I agree with both points then. If we can simplify the overall code I'd appreciate. Let's not overcomplicate it for the sake of it, if users are going to ask for more we can eventually expand, but I think the NTSC-J as it is might be the sweet spot everyone would be curious to play with.

> Also, you may recall that we had to take responsibility for movie tv->pc range expansion away from ffmpeg and do it in FFNx's shader

If that's the case then I'm fine with it, as with ffmpeg I can try to pass the GPU texture that already exists created by FFMpeg into bgfx without having to download it on the CPU and reupload it back to the GPU. By doing so we get what we all know as Hardware Acceleration, which means we decode the video data on the GPU and we keep it there. If the colors are fixed on the shader, they should apply as usual through the shader pipeline you already built.

Let me know if you want to handle all these changes into this PR or if you prefer to open a new one ( like if the code we're going to merge on this one is going to be deleted on the next one, there's no point in merging ).

Cheers :)

ChthonVII commented:
I apologize for not noticing your reply sooner.

I'm trying hard to keep each PR limited to stuff that (a) works on its own, and (b) is not going to change in the next PR.

However, I think I want to keep messing with the black level a bit. Now that I have the whole kit and kaboodle working (I think), I can consider more of the game when picking the black level. Also, I'd mucked up git when I was testing before, so I'm not sure the tests that led to the current choice were valid.

Hopefully I should have everything sorted in a few days.

ChthonVII commented:
Rebased for the past few upstream commits.

The next two pull requests are ready to go, as soon as I have time to write out the "summary," "motivation," etc. If you want a sneak peek, they are [here](https://github.com/ChthonVII/FFNx/tree/CRTgammaforeveryone) and [here](https://github.com/ChthonVII/FFNx/tree/reviseNTSCJmode). I hope to write those up tomorrow. And then do one more pull removing the per-asset gamut override stubs. 

After that, it might be time to think about the possibility of removing uncorrected mode in the name of code simplification.

julianxhokaxhiu reviewed (changes requested):
CI checks do not pass. I'll be happy to review once the CI goes green. Thank you.

//EDIT: Re-running them just in case but I'd appreciate if you could ping me and let me know if the CI fails because of reasons outside of your own control.

ChthonVII commented:
> CI checks do not pass. I'll be happy to review once the CI goes green. Thank you.
> 
> //EDIT: Re-running them just in case but I'd appreciate if you could ping me and let me know if the CI fails because of reasons outside of your own control.

All the failed checks yesterday were due to vcpkg failing to download stuff. I did note this over on pull #825.

That said, I'm considering messing with the gamma constants some more. So please hold off from merging for a bit.

julianxhokaxhiu commented:
Ok, CI seems to pass now, it was just Github runners taking too long to build. Vcpkg is just fine :) I'll hold on then, lmk when ready for review. Thank you

ChthonVII commented:
I think I'm just going to do one big post here, and then link to it in the other pull requests. I think everything's ready to go at this point. I have a few major thoughts I want to throw out here:

1. Starting with the embarrassing one: I've learned the hard way that when Visual Studio/cmake copies files into `.build/bin/` it does ***NOT*** overwrite existing files. I screwed myself up so many times when I tested with the wrong binary or wrong LUT files and ended up thinking something wasn't working when in fact it was. In two cases I reverted something and then put it back again; this is why. (This is probably also how I botched the diagnosis of #818. Twice.)
2. I wrote up [some pretty substantial documentation](https://github.com/ChthonVII/FFNx/blob/reviseNTSCJmode/docs/gamma_and_gamut.md) for the gamma and gamut stuff. I hope this proves worth reading and gives a useful explanation of what this code is doing, and what's baked into those mystery LUTs, and why.
3. I'm trying to break up the pull requests into digestible chunks at points where FFNx should be in a correct working state after each merge. But I think maybe I've hit the limit of how far I can run ahead of stuff getting merged while keeping things straight and orderly.
4. It turns out that I was able to prevent BGFX from doing sRGB linearization in the first place. (This is one of those instances where my changes worked the first time, but I tested with the wrong binary.) It looks to me like it's working with no unintended side effects. But this is a pretty big change, so I'd appreciate a second opinion.
5. Likewise, I can't test HDR or FF8, but both really need tested. (In particular the revised NTSC-J mode needs tested in HDR.)
6. Just FYI: I wanted to do 128x128x128 LUTs, but BGFX crapped out on them. So we're back to 64x64x64. I did try to stretch their bandwidth by storing values in gamma space.
7. What's next? Three things come to mind:
     - Maybe remove uncorrected mode and make everything NTSC-J mode all the time? 
          - The main reason to consider this is code simplification. And possibly slight performance gain.
          - The main reason not to is that modding is easier in uncorrected mode, where what you see in Photoshop, etc. is what you get in FFNx. But...
               - I already broke "what you see is what you get" a little bit when I changed the gamma.
               - The revised NTSC-J mode is closer to uncorrected mode. So usually stuff created in sRGB looks "good enough."
               - It's easier now to make mods for NTSC-J mode using sRGB tools like Photoshop, etc. There is now a Windows binary for gamutthingy, and [instructions](https://github.com/ChthonVII/FFNx/blob/reviseNTSCJmode/docs/mods/preparing_textures_for_ntscj_mode.md) for what to do with it to make mods for NTSC-J mode.
          - I'm pretty biased here. So I'd like to know what other people think after trying out the revised NTSC-J mode. Would they ever go back to uncorrected mode?
     -  The issue in that "substantial documentation" captioned "A Minor Shortcoming" bothers me way more than it should. I'd like to fix it, and I think I see how. But... it would be an unholy mess of conditionals, unless uncorrected mode was removed first.
     - [zaphod77](https://github.com/zaphod77) wanted a feature to disable dithering. (#757) So... On the one hand, the dithering is *correct*, and zaphod's problem is due to viewing a low-DPI screen closer than its minimum recommended viewing distance, and adding a feature means complicating the code. On the other hand, it should be fairly easy to implement, and would make the user happy. Thoughts?

ChthonVII commented:
My refactor efforts are running into a problem with eye textures and battle background floors. If both advanced lighting and NTSC-J mode are enabled, these textures turn black. The problem goes away if I stop color correcting textures. Is there anything special about these textures? Like are they maps or something?

ChthonVII commented:
#832 incorporates these changes and supersedes my other pull requests. I'm leaving this one outstanding because it could stand on its own.

julianxhokaxhiu commented:
Just to clarify, does #832 superseed this one? If yes, please close this one and rebase the other one so we can get it merged it cleanly. Thank you!

ChthonVII commented:
> Just to clarify, does #832 superseed this one? If yes, please close this one and rebase the other one so we can get it merged it cleanly. Thank you!

#832 includes this one (#816). This one (#816) exists as an an independent pull request because you **_might_** hypothetically want to just fix the movie gamma independently before undertaking review of the whole NTSCJ overhaul. If you're not inclined to do that, you certainly can just close this one (#816) and merge #832.

I'll try to find time to rebase #832 tomorrow.
